Implement, test, and benchmark new proposals

conectado opened the pull request:
# This PR addresses #1748 and #1864

In this PR I did the following things:
* For existing proposals updated the names to closely match those in the handbook
  * Text -> Signal
  * Spending -> Funding Request
  * Set Validator Count -> Set Max Validator Count
  * Add Working Group Leader Opening -> Create Working Group Lead Opening
  * Fill Working Group Leader Opening -> Fill Working Group Lead Opening
  * Set Working Group Budget Capacity ->  Update Working Group Budget
  * Decrease Working Group Leader Stake -> Decrease Working Group Lead Stake
  * Slash Working Group Leader Stake -> Slash Working Group Lead
  * Set Working Group Leader Reward -> Set Working Group Lead Reward
  * Terminate Working Group Leader Role -> Terminate Working Group Lead
* Removed old proposals no longer in the handbook
* Added new proposals:
  * Cancel Working Group Lead Opening
  * Set Membership Price
  * Set Council Budget Increment
  * Set Councilor Reward
  * Set Initial Invitation Balance
  * Set Initial Invitation Count
  * Set Membership Lead Invitation Quota
  * Set Referral Cut
  * **Note:** _Update Budget Request_ and _Funding Request_ changed significantly
* Needed to change the `Balance` type in `Council` to the `Balance` in the `balances` pallet to be able to update the budget of `WorkingGroup` with the budget from `Council`
* Updated `councilor_reward` and `budget_incrment` to be part of the storage as to be able to be updated with `Set Council Budget Increment` and `Set Councilor Reward`

### Also, I have a couple of open questions:
* In [`Set Max Validator Count`](https://joystream.gitbook.io/joystream-handbook/governance/proposals#set-max-validator-count) we use `<staking::Module<T>>::minimum_validator_count()` instead of `MIN_VALIDATOR_COUNT`. Is this okay?


### [Handbook companion PR](https://github.com/Joystream/handbook/pull/27)

conectado committed:
runtime: proposal/codex: move proposal creation parameters into struct

conectado committed:
runtime: proposals: add set membership price and cancel working group lead

* Adds new proposals
* Add additional checks for funding request

conectado committed:
runtime: proposals: modify existing proposal match handbook

* Modifies existing proposal name, parameters, and check to match handbook
* Add structure to create Update budget proposal

conectado committed:
runtime: proposals: update tests to match new proposal names

conectado committed:
runtime: proposals/codex: update tests with checks changes

conectado committed:
runtime: proposals: fix compilation after rebase

conectado committed:
runtime: proposal: add structure for new proposals

Adds neccesary infrastructure with subs for:
* Set council budget increment
* Set Councilor Reward
* Set Initial Invitation Balance

conectado committed:
benchmark: proposals/codex: add benchmarks and weights

conectado committed:
benchmark: proposal: add weight runtime_upgrade and comments

conectado committed:
runtime: proposals: fix tests

conectado committed:
runtime: proposals: fix old proposals and add new

Fixed:
* `FundingRequest`
* `UpdateWorkingGroupBudget`

Added:
* `SetInitialInvitationBalance`
* `SetInitialInvitationCount`
* `SetMembershipLeadInvitationQuota`
* `SetReferralCut`

conectado committed:
runtime: proposals/codex: update council and add set council budget increment

* Update council to use budget increment as a mutable storage
* Add set council budget increment proposal

conectado committed:
runtime: proposals: fix council proposals

This fixes the proposals for:
* Set Council Budget Increment
* Set Councilor Reward

Furthermore this neccesitated updating council for having councilor_reward
and budget increment to be stored as part of the module's storage

conectado committed:
proposals: Add weights benchmarks and tests for missing proposals

conectado committed:
Fix errors after rebase

conectado committed:
runtime: working-group: move instances aliases to working-group crate

shamil-gadelshin commented on runtime-modules/proposals/codex/src/lib.rs:
```diff
@@ -58,40 +58,92 @@ mod types;
     + common::Trait
+    + council::Trait
     + staking::Trait
+    + working_group::Trait<ForumWorkingGroupInstance>
```
This will require the codex update for each new working group created.

conectado replied:
I could fix this by adding a hook to the runtime that allows us to query and set the working group budget. Also, this would fix having the working group instances definitions in the working group since this way I would only need them in the runtime and I can move them back there. Do you think this is an acceptable way of handling this?

shamil-gadelshin replied:
Yes, I also thought of the runtime hook. Let's create a prototype to see how it goes.

shamil-gadelshin commented on runtime/src/proposals_configuration/sample_proposal_parameters.json:
```diff
@@ -1,24 +1,25 @@
-  "amend_constitution_proposal": {}
-}
\ No newline at end of file
+    "_comment": "This is a sample 'Proposal parameters' JSON file to enable conditional runtime compilation.",
```
Please, modify this file to reflect your changes - it is needed by the integration tests.

shamil-gadelshin replied:
Make sure that this file works, you can use this `proposal_parameters_are_initialized` test for each proposal locally (but don't create tests for each proposal - we won't enable it by default because of the conditional compilation).

shamil-gadelshin commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -63,6 +63,18 @@ use staking_handler::StakingHandler;
 
 type WeightInfoWorkingGroup<T, I> = <T as Trait<I>>::WeightInfo;
 
+/// The forum working group instance alias.
```
You made the working group - instance-dependent which downgrades its flexibility. For each new working group - you should update it.

shamil-gadelshin commented on runtime-modules/proposals/codex/src/types.rs:
```diff
@@ -96,43 +136,43 @@ pub struct GeneralProposalParams<MemberId, AccountId, BlockNumber> {
 
-    /// Terminate role slash penalty.
-    pub penalty: Option<Penalty<Balance>>,
+    /// Optional amount to be slashed.
```
Why did you remove the penalty here? My guess because the gitbook doesn't contain it. But termination slash is similar to the regular slash. And it contains the slashing rationale. You could confirm it with Bedeho.

shamil-gadelshin replied:
I think that the gitbook should be updated rather than changing this parameter.

conectado replied:
I think this parameter is missing in the gitbook because the generic description for the proposal already doubles as a rationale. I remember bedeho saying he dropped the rationale from slashing lead for this reason.

shamil-gadelshin replied:
Ok. Make a comment please when you pass the empty text on two occasions: slashing and termination.

shamil-gadelshin commented on runtime-modules/proposals/codex/src/benchmarking.rs:
```diff
@@ -0,0 +1,811 @@
+    }
+
+    update_working_group_budget_positive_forum {
+        set_wg_and_council_budget::<T>(100, WorkingGroup::Forum);
```
Is it possible to introduce some parameter x=0..1 which could define the balance sign? This could reduce the benchmarks count and simplify the code.

conectado replied:
We could do that, but I am a little reticent to do it since the parameter would imply an order from best to worst-case scenario and the benchmarks will try to do a linear regression over it and might create and spurious linear dependency.

shamil-gadelshin replied:
Fair enough. Let it be.

shamil-gadelshin commented on runtime-modules/proposals/codex/src/lib.rs:
```diff
@@ -372,31 +538,87 @@ decl_module! {
+                    ensure!(amount <= wg_budget, Error::<T>::InsufficientFundsForFundingRequest);
+
+                    call_wg!(working_group<T>, set_budget, origin.clone(), wg_budget - amount)?;
+                    Council::<T>::set_budget(origin, current_budget.saturating_add(amount))?;
```
Why does it set the council budget rather than working group budget?

conectado replied:
Since the balance is non-negative in the line before this I reduce the working group budget by `amount` and in this line I add that `amount` to the council budget. That's the behavior I understand is expected [from the handbook](https://joystream.gitbook.io/joystream-handbook/governance/proposals#update-working-group-budget).

shamil-gadelshin replied:
Sorry, I missed the call_wg part.

shamil-gadelshin commented on runtime-modules/proposals/codex/src/lib.rs:
```diff
@@ -372,31 +538,87 @@ decl_module! {
+
+        /// Funding request proposal
+        #[weight = WeightInfoCodex::<T>::funding_request()] // TODO: adjust weight
+        pub fn funding_request(
```
You don't need this extrinsic. Because you could just run the extrinsic from the council module.

shamil-gadelshin commented on runtime-modules/proposals/codex/src/lib.rs:
```diff
@@ -372,31 +538,87 @@ decl_module! {
                 ensure!(
-                    *balance != BalanceOf::<T>::zero(),
-                    Error::<T>::InvalidSpendingProposalBalance
+                    *amount <= <BalanceOf<T>>::from(MAX_SPENDING_PROPOSAL_VALUE),
```
We don't check for the storage of other modules when create the proposal. We check only proposal parameters. Ask Bedeho for details.

conectado replied:
I see, in the [handbook](https://joystream.gitbook.io/joystream-handbook/governance/proposals#creation-conditions-3) it says that it should check for council balance, so I will ask for clarification

shamil-gadelshin reviewed (changes requested):
Great work, but there are major issues:
- working group dependency on its own instances
- codex dependency on the working group instances
- unnecessary extrinsic proxies

Also, please, update the codex pallet comment.

conectado committed:
proposal: move funding request extrinsic to council

conectado committed:
proposals: integration: update test json

conectado committed:
proposals: add claryfing comments for slashing_text

conectado committed:
runtime: proposals/codex: decouple proposal codex trait from workgin-group

By adding a hook to the runtime we decouple the trait itself with the proposal
codex. And with this we don't need to update the trait for proposal codex for
each instance of the working group.

conectado committed:
runtime: proposals/codex: update weights

conectado committed:
runtime: proposal/codex: update comment

conectado committed:
runtime: proposal/codex: update SetMaxValidatorCount creation comment

conectado committed:
runtime: working-group: eliminate instance dependency

conectado committed:
council: tests: fix after rebase

conectado committed:
runtime: codex: improvement to runtime hooks

shamil-gadelshin commented on runtime/src/proposals_configuration/tests.rs:
```diff
@@ -1,10 +1,371 @@
+#[ignore]
+fn proposal_parameters_are_initialized_max_validators() {
+    let actual_params = super::SetMaxValidatorCountProposalParameters::get();
+    let expected_params = ProposalParameters {
```
Please, consider creating "default_proposal_parameters" function.

shamil-gadelshin commented on runtime-modules/working-group/src/lib.rs:
```diff
@@ -1029,6 +1029,16 @@ decl_module! {
 
 impl<T: Trait<I>, I: Instance> Module<T, I> {
+    // Set working group budget
+    pub fn put_budget(new_budget: BalanceOf<T>) {
```
Please, remove these duplicated methods.

shamil-gadelshin reviewed (changes requested):
There are still a couple of minor issues.

conectado committed:
runtime: working-group: remove duplicated methods

shamil-gadelshin reviewed (approved):
LGTM
